src/matrix_common/types/mxc_uri.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2022 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Type, TypeVar
+from urllib.parse import urlparse
+
+import attr
+
+MU = TypeVar("MU", bound="MXCUri")
+
+
+@attr.s(frozen=True, slots=True, auto_attribs=True)
+class MXCUri:
+    """Represents a URI that points to a media resource in matrix.
+
+    MXC URIs take the form 'mxc://server_name/media_id'.
+    """
+
+    server_name: str
+    media_id: str
+
+    @classmethod
+    def from_str(cls: Type[MU], mxc_uri_str: str) -> MU:
+        """
+        Given a str in the form "mxc://<domain>/<media_id>", return an equivalent MXCUri.
+
+        Args:
+            mxc_uri_str: The MXC Uri as a str.
+
+        Returns:
+            An MXCUri object with matching attributes.
+
+        Raises:
+            ValueError: If the str was not a valid MXC Uri.
+        """
+        # Attempt to parse the given URI. This will raise a ValueError if the uri is
+        # particularly malformed.
+        parsed_mxc_uri = urlparse(mxc_uri_str)
+
+        # MXC Uri's are pretty bare bones. The scheme must be "mxc", and we don't allow
+        # any fragments, query parameters or other features.
+        if (
+            # The scheme must be "mxc".
+            parsed_mxc_uri.scheme != "mxc"
+            # There must be a host and path provided.
+            or not parsed_mxc_uri.netloc
+            or not parsed_mxc_uri.path
+            or not parsed_mxc_uri.path.startswith("/")
```

Comment:
Can the `path` ever start with anything besides `/`?

Reply:
Looks like it can if you don't include a scheme (i.e. `domain.com/media_id`). But then, it's not a valid [MXC URI](https://spec.matrix.org/v1.3/client-server-api/#matrix-content-mxc-uris).

Reply:
This parses into:

```pycon
>>> urlparse("domain.com/media_id")
ParseResult(scheme='', netloc='', path='domain.com/media_id', params='', query='', fragment='')
```

Which is rather surprising!

But in this case the `scheme` wouldn't match... well it seems fine regardless. 👍